clpy/core/core.pyx

```diff
@@ -2979,17 +2979,23 @@ cpdef _scatter_op_single(ndarray a, ndarray indices, v,
         #     v, indices, cdim, rdim, adim, a.reduced_view())
         raise NotImplementedError("clpy does not support this")
     elif op == 'add':
-        # There is constraints on types because atomicAdd() in CUDA 7.5
-        # only supports int32, uint32, uint64, and float32.
-        # if not issubclass(v.dtype.type,
-        #                   (numpy.int32, numpy.float32,
-        #                    numpy.uint32, numpy.uint64, numpy.ulonglong)):
-        #     raise TypeError(
-        #         'scatter_add only supports int32, float32, uint32, uint64 as'
-        #         ' data type')
-        # TODO(vorj): fix this after resolve #30
-        if not issubclass(v.dtype.type, numpy.float32):
-            raise NotImplementedError('clpy does not support this')
+        # NOTE(clpy):
+        # CuPy had supported
+        # int32, float32, uint32, uint64, numpy.ulonglong.
+        # OpenCL without extension can't perform 64-bit
+        # atomic operations.
+        # Thus we support int32, float32, and uint32.
+        if issubclass(v.dtype.type, numpy.uint64):
+            raise NotImplementedError(
+                'OpenCL without extension'
```

Comment:
This message could make users misunderstanding supported types. They could be think that ClPy supports uint64 atomic op if the device has ` cl_khr_int64_*_atomics`.

I propose removing this and merge the below message `only supports int32, float32, uint32`

Reply:
And `only supports...` messages is on only ClPy. Please write it down.

Reply:
OK. I wrote explicitly the difference of the functionality of cupy's and clpy's `scatter_add` on the error message.